### ISSUE TITLE
[macos] Cannot translate from mac keycode: (12,568)

### ISSUE THREAD

**Issue by Alounv** [kmonad/kmonad#817]
I have this error message from time to time since I installed my new `Unicorne` split keyboard (it's a Corne basically).
When it happens kmonad shuts down.
```
Encountered error in KeySource: [Cannot translate from mac keycode: (12,568)]
kmonad: [Cannot translate from mac keycode: (12,568)]
```
Any idea what could be wrong, could we just ignore the error instead of crashing the whole program?

**Comment by Erlendms** [comment]
I have "the same" issue:
```
Encountered error in KeySource: [Cannot interpret mac switch: 205,Cannot translate from mac keycode: (12,4294967295)]
kmonad: [Cannot interpret mac switch: 205,Cannot translate from mac keycode: (12,4294967295)]
```

**Comment by Alounv** [comment]
@Erlendms for information, my workaround is to wrapp the kmonad launch in a retry inside my .zshrc (if you use zsh).

```
retry() {
    local retries=20
    local delay=1
    local count=0

    while [[ $count -lt $retries ]]; do
        if "$@"; then
            return 0
        fi
        count=$((count + 1))
        sleep $delay
    done

    return 1
}

alias kmo="retry sudo kmonad ~/.config/kmonad/aloun.kbd"


```

**Comment by gmtborges** [comment]
I fixed it using Kanata 👀

I only installed it with `brew install kanata` and remove the defcfg part of my .kbd, besides now I need only one config to my external and mac keyboard.

**Comment by jokesper** [comment]
Do you know how Kanata differentiates between devices it should and shouldn't work on?

**Comment by gmtborges** [comment]
> Do you know how Kanata differentiates between devices it should and shouldn't work on?

There is a config for mac and linux, here an example:
```
(defcfg
  macos-dev-names-include (
    "Device name 1"
    "Device name 2"
  )
  linux-dev-names-include ("Name 1" "Name 2")
)
```

**Comment by jokesper** [comment]
So it doesn't do autodetection differently and just allows multiple devices (kind of like #596)

**Comment by gmtborges** [comment]
> So it doesn't do autodetection differently and just allows multiple devices (kind of like [#596](https://github.com/kmonad/kmonad/pull/596))

Defining the inputs are optional, by default kanata listen to all of them. So to configure home row mod I only need one config file for my external and mac keyboard

```
(defcfg
  process-unmapped-keys yes
  concurrent-tap-hold   yes
)

(defsrc
  caps a    s    d    f    g    h    j    k    l    ;    '    ret
)

(defalias
    met_a (tap-hold-release 200 200 a lmet)
    alt_s (tap-hold-release 200 200 s lalt)
    ctl_d (tap-hold-release 200 200 d lctl)
    sft_f (tap-hold-release 200 200 f lsft)

    sft_j (tap-hold-release 200 200 j rsft)
    ctl_k (tap-hold-release 200 200 k rctl)
    alt_l (tap-hold-release 200 200 l lalt)
    met_; (tap-hold-release 200 200 ; rmet)
)

(deflayer base
  _ @met_a @alt_s @ctl_d @sft_f g h @sft_j @ctl_k @alt_l @met_; _  _
)

```

**Comment by jokesper** [comment]
I think @jsimonetti case is that, their mouse is detected as a keyboard.
KMonad too allows (under mac) to listen to all keyboards. The problem is the invalid detection. Either we need to detect better or we need to fallthrough those events without converting them. Ideally both.
The problem for the later is that it's rather complicated and would be a slightly bigger change.

**Comment by jsimonetti** [comment]
This is correct. I believe, because the mouse has back/forward buttons it is (also) registering as a keyboard.
If we could modify kmonad to also allow binding to (multiple) explicitly specified keyboard, this problem would go away.

If, in my case, I could bind to just my internal and external keyboard, mouse events (and mouse buttons) would not be caught by kmonad altogether. So it would make sense to invest into binding to multiple devices at a time.

(an alternative, already mentioned elsewhere, is to just run kmonad once for each keyboard to want to listen to)

**Comment by plasmadice** [comment]
I'm getting this error as soon as I touch my mouse. Has to be mouse related because I had a similar issue in Kanata that only went away after using `macos-dev-names-include` to exclude that device, which introduced me to [another issue](https://github.com/jtroo/kanata/issues/1539).

I'm using a Logitech M720 mouse. 

In Kanata, the mouse would get stuck in the top-left corner. In Kmonad, it just exits with OPs error.
